scio-core/src/main/scala/com/spotify/scio/io/TextIO.scala

```diff
@@ -45,32 +47,68 @@ final case class TextIO(path: String) extends ScioIO[String] {
         .withCompression(params.compression)
     )
 
+  private def textOut(
+    write: BTextIO.Write,
+    path: String,
+    suffix: String,
+    numShards: Int,
+    compression: Compression,
+    header: Option[String],
+    footer: Option[String],
+    shardNameTemplate: String,
+    tempDirectory: ResourceId,
+    filenamePolicyCreator: FilenamePolicyCreator,
+    isWindowed: Boolean
+  ) = {
+    if (tempDirectory == null) throw new IllegalArgumentException("tempDirectory must not be null")
+    if (shardNameTemplate != null && filenamePolicyCreator != null)
+      throw new IllegalArgumentException(
+        "shardNameTemplate and filenamePolicyCreator may not be used together"
+      )
+
+    val fp = Option(filenamePolicyCreator)
+      .map(c => c.apply(ScioUtil.pathWithPrefix(path, ""), suffix))
+      .getOrElse(
+        ScioUtil.defaultFilenamePolicy(
```

Comment:
Since this logic block seems to get repeated for each IO, what if we added more of it to the `ScioUtil.defaultFilenamePolicy` helper? Something like 

```scala
object ScioUtil {
  def resolveFilenamePolicy(
    policy: Option[FilenamePolicyCreator],
    path: String,
    suffix: String,
    shardNameTemplate: String,
    isWindowed: Boolean
  ): FilenamePolicyCreator =
    Option(filenamePolicyCreator)
      .map(c => c.apply(pathWithPrefix(path, ""), suffix))
      .getOrElse(
        defaultFilenamePolicy(pathWithPrefix(path),
          shardNameTemplate,
          suffix,
          isWindowed
        )
      )
```

Reply:
Sadly this isn't exactly the same in all cases to maintain some semblance of backwards compatibility

Reply:
Looking at this again, was able to simplify